vochain/state.go

```diff
@@ -78,65 +212,115 @@ type ImmutableState struct {
+	}
+	version, err := sdb.Version()
+	if err != nil {
+		return nil, err
+	}
+	root, err := sdb.Hash()
 	if err != nil {
 		return nil, err
 	}
 	log.Infof("state database is ready at version %d with hash %x",
-		vs.Store.Version(), vs.Store.Hash())
-	return vs, nil
+		version, root)
+	tx, err := sdb.BeginTx()
+	if err != nil {
+		return nil, err
+	}
+	mainTreeView, err := sdb.TreeView(nil)
+	if err != nil {
+		return nil, err
+	}
+	s := &State{
+		Store:     sdb,
+		Tx:        tx,
+		voteCache: voteCache,
+	}
+	s.setMainTreeView(mainTreeView)
+	return s, nil
 }
 
-// initStore initializes the state iavl tree
-func initStore(dataDir string, state *State) error {
-	log.Infof("initializing state db store")
-	state.Store = new(iavlstate.IavlState)
-	// state.Store = new(gravitonstate.GravitonState)
-	if err := state.Store.Init(dataDir, "disk"); err != nil {
-		return err
+// initStateDB initializes the StateDB with the default subTrees
+func initStateDB(dataDir string) (*statedb.StateDB, error) {
+	log.Infof("initializing StateDB")
+	db, err := badgerdb.New(badgerdb.Options{Path: dataDir})
+	if err != nil {
+		return nil, err
 	}
+	sdb := statedb.NewStateDB(db)
 	startTime := time.Now()
-	if err := state.Store.AddTree(AppTree); err != nil {
-		return err
+	defer log.Infof("StateDB load took %s", time.Since(startTime))
+	root, err := sdb.Hash()
+	if err != nil {
+		return nil, err
 	}
-	if err := state.Store.AddTree(ProcessTree); err != nil {
-		return err
+	if !bytes.Equal(root, make([]byte, len(root))) {
+		// StateDB already initialized if StateDB.Root != emptyHash
+		return sdb, nil
 	}
-	if err := state.Store.AddTree(VoteTree); err != nil {
-		return err
+	update, err := sdb.BeginTx()
+	defer update.Discard()
+	if err != nil {
+		return nil, err
 	}
-	//	if err := state.Store.AddTree(CensusTree); err != nil {
-	//		return err
-	//	}
-	log.Infof("state tree load took %s", time.Since(startTime))
-	return nil
+	if err := update.Add(OraclesCfg.Key(),
+		make([]byte, OraclesCfg.HashFunc().Len())); err != nil {
+		return nil, err
+	}
+	if err := update.Add(ValidatorsCfg.Key(),
+		make([]byte, ValidatorsCfg.HashFunc().Len())); err != nil {
+		return nil, err
+	}
+	if err := update.Add(ProcessesCfg.Key(),
+		make([]byte, ProcessesCfg.HashFunc().Len())); err != nil {
+		return nil, err
+	}
+	header := models.TendermintHeader{
+		Height:  0,
+		AppHash: []byte{},
+		ChainId: "empty",
+	}
+	headerBytes, err := proto.Marshal(&header)
+	if err != nil {
+		return nil, err
+	}
+	if err := update.Add(headerKey, headerBytes); err != nil {
+		return nil, err
+	}
+	return sdb, update.Commit()
+}
+
+// mainTreeView is a thread-safe function to obtain a pointer to the last
+// opened mainTree as a TreeView.
+func (v *State) mainTreeView() *statedb.TreeView {
+	return v.mainTreeViewValue.Load().(*statedb.TreeView)
+}
+
+// TODO: @mvdan is the usage of atomic.Value appropiate here?  Or should I
+// better use atomic.LoadPointer & atomic.StorePointer?
```

Comment:
your use of atomic.Value is correct, I think. if you used pointers, you would be forced into using unsafe, which is... less safe. If you mess up with `interface{}`, you will get a helpful runtime panic (in the line 303 above, when you type assert). If you mess up with `unsafe.Pointer`, you may end up with memory corruption or worse.